src/kOS/UserIO/TelnetSingletonServer.cs

```diff
@@ -0,0 +1,1064 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using kOS.Safe.Utilities;
+using kOS.Module;
+using UnityEngine;
+using System.IO;
+using System.IO.Pipes;
+using kOS.Safe.UserIO;
+
+namespace kOS.UserIO
+{
+    /// <summary>
+    /// Spawned by TelnetMainServer, this handles the connection to just one
+    /// single telnet client, leaving the main server free to go back to listening
+    /// to other new connecting clients.
+    /// </summary>
+    public class TelnetSingletonServer
+    {
+        // ReSharper disable SuggestUseVarKeywordEvident
+        
+        // I want to communicate the difference between cases where the fact that the
+        // value starts at zero is irrelevant to the algorithm from cases where it's
+        // actually part of the algorithm:
+        // ReSharper disable RedundantDefaultFieldInitializer
+        
+        private volatile TcpClient client;
+        
+        private TelnetMainServer whoLaunchedMe;
+        
+        /// <summary>
+        /// The raw socket stream used to talk directly to the client across the network.
+        /// It is bidirectional - handling both the input from and output to the client.
+        /// </summary>
+        private NetworkStream rawStream;
+        
+        /// <summary>
+        /// The queue that other parts of KOS can use to read characters from the telnet client.
+        /// </summary>
+        private volatile Queue<char> inQueue;
+        private readonly object inQueueAccess = new object(); // To make all access of the inQueue atomic between threads.
+
+        /// <summary>
+        /// The queue that other parts of kOS can use to write characters to the telent client.
+        /// </summary>
+        private volatile Queue<char> outQueue;
+        private readonly object outQueueAccess = new object(); // To make all access of the outQueue atomic between threads.
+        
+        public kOSProcessor ConnectedProcessor { get; private set; }
+        
+        private Thread inThread;
+        private Thread outThread;
+        
+        private TelnetWelcomeMenu welcomeMenu;
+        
+        private TerminalUnicodeMapper terminalMapper;
+        
+        /// <summary>
+        /// If true, then all input should get consumed and ignored the next time inThread checks, until the input stream is empty again.
+        /// </summary>
+        private bool flushPendingInput;
+                        
+        /// <summary>
+        /// What order was I launched in?  Was I the 1st TelnetSingletonServer? The 2nd?  The 423rd?
+        /// </summary>
+        private int MySpawnOrder { get; set; }
+        
+        private bool isLineAtATime;
+        private bool allowResize;
```

Comment:
I had an idea of making one window slaved to the other so as to avoid the resize arguments between them but it turns out you cannot enforce that over telnet anyway.  If the client wants to resize itself then tough, you have to let it.  This can be removed.